controllers/machine.go

```diff
@@ -82,48 +115,58 @@ func (mrc *machineReconcileContext) addFinalizer() error {
 	return nil
 }
 
-func (mrc *machineReconcileContext) ensureDependencies() error {
-	hardware, err := mrc.ensureHardware()
-	if err != nil {
-		return fmt.Errorf("ensuring hardware: %w", err)
-	}
+func (mrc *machineReconcileContext) isHardwareReady(hw *tinkv1.Hardware) bool {
+	return hw.Spec.Metadata.State == inUse && hw.Spec.Metadata.Instance.State == provisioned
+}
 
-	if err := mrc.ensureTemplate(hardware); err != nil {
-		return fmt.Errorf("ensuring template: %w", err)
+func (mrc *machineReconcileContext) ensureTemplateAndWorkflow(hw *tinkv1.Hardware) error {
+	// create workflows and templates
+	if err := mrc.ensureTemplate(hw); err != nil {
+		return fmt.Errorf("failed to ensure template: %w", err)
 	}
 
-	if err := mrc.ensureWorkflow(hardware); err != nil {
-		return fmt.Errorf("ensuring workflow: %w", err)
+	if err := mrc.ensureWorkflow(hw); err != nil {
+		return fmt.Errorf("failed to ensure workflow: %w", err)
 	}
 
 	return nil
 }
 
-func (mrc *machineReconcileContext) markAsReady() error {
-	mrc.tinkerbellMachine.Status.Ready = true
+//
+func (mrc *machineReconcileContext) Reconcile() error {
+	defer func() {
+		// make sure we do not create orphaned objects.
+		if err := mrc.addFinalizer(); err != nil {
+			mrc.log.Error(err, "error adding finalizer")
+		}
+	}()
 
-	if err := mrc.patch(); err != nil {
-		return fmt.Errorf("patching machine with ready status: %w", err)
+	hw, err := mrc.ensureHardware()
+	if err != nil {
+		return fmt.Errorf("failed to ensure hardware: %w", err)
 	}
 
-	return nil
-}
+	if !mrc.isHardwareReady(hw) {
+		found, err := mrc.workflowExists()
+		if err != nil || !found {
```

Comment:
Would we not want to immediately return if there's an error?

Reply:
Ideally we don't have bool errors either. This could be modelled like `io.EOF` or `apierrors.IsNotFound()`. This would ensure we only retrieve the workflow once where we have a double call at the moment (not that big a deal considering it hits a cache).

```go
workflow, err := mrc.getWorkflow(...)
switch{
case apierrors.IsNotFound(err):
  // ensure template and workflow returns the workflow it creates.
  workflow, err = mrc.ensureTemplateAndworkflow(hw) ...
case err != nil:
  // something else happened so return error forcing a requeue.
}

if lastActionStarted(workflow) {
  // mark ready
}
```

Reply:
updated. mind taking a look?